File: java/src/main/java/com/languageweaver/sdk/samples/feedback/EdgeFeedbackService.java
package com.languageweaver.sdk.samples.feedback;

import com.languageweaver.sdk.common.constants.ApprovalStatus;
import com.languageweaver.sdk.common.constants.FeedbackComment;
import com.languageweaver.sdk.common.edge.EdgeLanguageWeaverClient;
import com.languageweaver.sdk.feedback.edge.request.*;
import com.languageweaver.sdk.feedback.edge.result.EdgeFeedbackListResult;
import com.languageweaver.sdk.feedback.edge.result.EdgeFeedbackResult;

import java.io.IOException;
import java.text.ParseException;
import java.util.List;

public class EdgeFeedbackService {

    public static EdgeFeedbackResult createFeedback() throws IOException {
        try (EdgeLanguageWeaverClient lwClient = new EdgeLanguageWeaverClient().build()) {
            EdgeCreateFeedbackRequest createFeedbackRequest = new EdgeCreateFeedbackRequest()
                    .setSourceLanguageId("eng")
                    .setTargetLanguageId("fra")
                    .setModel("generic")
                    .setSourceText("sourceText")
                    .setTargetText("targetText")
                    .setSuggestedTranslation("suggestedTranslation")
                    .setComment(FeedbackComment.SPELLING)
                    .setApprovalState(ApprovalStatus.APPROVED);

            EdgeFeedbackResult feedback = lwClient.createFeedback(createFeedbackRequest);
            return feedback;
        }
    }

    public static void updateFeedback() throws IOException {
        try (EdgeLanguageWeaverClient lwClient = new EdgeLanguageWeaverClient().build()) {
            EdgeUpdateFeedbackRequest edgeUpdateFeedbackRequest = new EdgeUpdateFeedbackRequest()
                    .setFeedbackId("feedbackId")
                    .setSuggestedTranslation("new suggested translation")
                    .setComment(FeedbackComment.CAPITALIZATION_PUNCTUATION);

            EdgeFeedbackResult feedback = lwClient.updateFeedback(edgeUpdateFeedbackRequest);
        }
    }

    public static void updateFeedbackStatus() throws IOException {
        try (EdgeLanguageWeaverClient lwClient = new EdgeLanguageWeaverClient().build()) {
            EdgeUpdateFeedbackApprovalRequest edgeUpdateApprovalRequest = new EdgeUpdateFeedbackApprovalRequest()
                    .setFeedbackId("feedbackId")
                    .setApprovalStatus(ApprovalStatus.REJECTED);

            EdgeFeedbackResult feedback = lwClient.updateFeedbackApproval(edgeUpdateApprovalRequest);
            System.out.println(feedback.getApprovalState());
        }
    }

    public static void deleteFeedback() throws IOException {
        try (EdgeLanguageWeaverClient lwClient = new EdgeLanguageWeaverClient().build()) {
            EdgeDeleteFeedbackRequest deleteFeedbackRequest = new EdgeDeleteFeedbackRequest()
                    .setFeedbackId("feedbackId");

            lwClient.deleteFeedback(deleteFeedbackRequest);
        }
    }

    public static void retrieveFeedback() throws IOException {
        try (EdgeLanguageWeaverClient lwClient = new EdgeLanguageWeaverClient().build()) {
            int pageNumber = 1;
            int pageSize = 50;
            EdgeFeedbackListResult edgeFeedbackList;
            do {
                edgeFeedbackList = lwClient.getEdgeFeedback(pageNumber, pageSize);
                // process feedback
                List<EdgeFeedbackResult> feedbackList = edgeFeedbackList.getFeedbackList();
                feedbackList.stream()
                        .limit(3)
                        .forEach(edgeFeedbackResult -> System.out.println(edgeFeedbackResult.getFeedbackId()));
                pageNumber++;
            } while (edgeFeedbackList.getFeedbackList().size() == pageSize);
        }
    }

    public static void filterFeedback() throws IOException {
        try (EdgeLanguageWeaverClient lwClient = new EdgeLanguageWeaverClient().build()) {
            EdgeFilterFeedbackRequest filterFeedbackRequest = new EdgeFilterFeedbackRequest()
                    .setLanguagePairId("languagePairId")
                    .setSourceText("sourceText")
                    .setMachineTranslation("machineTranslation")
                    .setSuggestedTranslation("suggestedTranslation")
                    .setComment("comment")
                    .setReviewer("reviewer username")
                    .setUser("creation username")
                    .addApprovalStatus(ApprovalStatus.APPROVED)
                    .addApprovalStatus(ApprovalStatus.PENDING)
                    .setStartDate("10/10/2021")
                    .setEndDate("10/10/2022");

            int pageNumber = 1;
            EdgeFeedbackListResult filteredFeedback;
            do {
                filteredFeedback = lwClient.getEdgeFeedback(filterFeedbackRequest, pageNumber);
                // process feedback
                List<EdgeFeedbackResult> feedbackList = filteredFeedback.getFeedbackList();
                feedbackList.stream()
                        .limit(3)
                        .forEach(edgeFeedbackResult -> System.out.println(edgeFeedbackResult.getFeedbackId()));
                pageNumber++;
            } while (filteredFeedback.getFeedbackList().size() > 0);
        }
    }

    public static void main(String[] args) throws IOException, ParseException {
        try (EdgeLanguageWeaverClient lwClient = new EdgeLanguageWeaverClient().build()) {
            EdgeFeedbackResult feedback;

//          CREATE FEEDBACK
            feedback = createFeedback();
            System.out.println(feedback.getFeedbackId());

//          UPDATE FEEDBACK
            EdgeUpdateFeedbackRequest edgeUpdateFeedbackRequest = feedback.toUpdateRequest()
                    .setSuggestedTranslation("new suggested translation")
                    .setComment(FeedbackComment.CAPITALIZATION_PUNCTUATION);
            feedback = lwClient.updateFeedback(edgeUpdateFeedbackRequest);

//          UPDATE FEEDBACK APPROVAL
            EdgeUpdateFeedbackApprovalRequest edgeUpdateApprovalRequest = feedback.toUpdateApprovalRequest(ApprovalStatus.REJECTED);
            feedback = lwClient.updateFeedbackApproval(edgeUpdateApprovalRequest);
            System.out.println(feedback.getApprovalState());

//          DELETE FEEDBACK
            EdgeDeleteFeedbackRequest deleteFeedbackRequest = feedback.toDeleteRequest();
            lwClient.deleteFeedback(deleteFeedbackRequest);
        }
    }
}

File: dotnet/Samples/Feedback/EdgeFeedbackService.cs
using LanguageWeaver.Sdk.Common.Edge;
using LanguageWeaver.Sdk.Constants;
using LanguageWeaver.Sdk.Feedback.Edge.Request;
using LanguageWeaver.Sdk.Feedback.Edge.Result;

namespace LanguageWeaver.Sdk.Samples.Feedback;

public class EdgeFeedbackService
{
    public static void Main()
    {
        using var lwClient = new EdgeLanguageWeaverClient().Build();

        EdgeFeedbackResult feedback;

        //CREATE FEEDBACK
        feedback = CreateFeedback();
        Console.WriteLine(feedback.FeedbackId);

        //UPDATE FEEDBACK
        EdgeUpdateFeedbackRequest edgeUpdateFeedbackRequest = feedback.ToUpdateRequest();
        edgeUpdateFeedbackRequest.SuggestedTranslation = "new suggested translation";
        feedback = lwClient.UpdateFeedback(edgeUpdateFeedbackRequest);

        //UPDATE FEEDBACK APPROVAL
        EdgeUpdateFeedbackApprovalRequest edgeUpdateFeedbackApprovalRequest =
            feedback.ToUpdateApprovalRequest(ApprovalStatus.Rejected);
        feedback = lwClient.UpdateFeedbackApproval(edgeUpdateFeedbackApprovalRequest);
        Console.WriteLine(feedback.ApprovalState);

        //DELETE FEEDBACK
        EdgeDeleteFeedbackRequest deleteFeedbackRequest = feedback.ToDeleteRequest();
        lwClient.DeleteFeedback(deleteFeedbackRequest);
    }

    public static EdgeFeedbackResult CreateFeedback()
    {
        using var lwClient = new EdgeLanguageWeaverClient().Build();

        EdgeCreateFeedbackRequest edgeCreateFeedbackRequest = new EdgeCreateFeedbackRequest()
        {
            LanguagePairId = "languagePairId",
            SourceLanguageId = "eng",
            TargetLanguageId = "fra",
            Model = "generic",
            SourceText = "sourceText",
            TargetText = "targetText",
            SuggestedTranslation = "suggestedTranslation2",
            ApprovalState = ApprovalStatus.Approved
        };

        EdgeFeedbackResult feedback = lwClient.CreateFeedback(edgeCreateFeedbackRequest);
        return feedback;
    }

    public static void UpdateFeedback()
    {
        using var lwClient = new EdgeLanguageWeaverClient().Build();

        EdgeUpdateFeedbackRequest edgeUpdateFeedbackRequest = new EdgeUpdateFeedbackRequest()
        {
            FeedbackId = "feedbackId",
            SuggestedTranslation = "new suggested translation"
        };

        EdgeFeedbackResult feedback = lwClient.UpdateFeedback(edgeUpdateFeedbackRequest);

        Console.WriteLine(feedback.ApprovalState);
    }

    public static void UpdateFeedbackStatus()
    {
        using var lwClient = new EdgeLanguageWeaverClient().Build();

        EdgeUpdateFeedbackApprovalRequest edgeUpdateFeedbackApprovalRequest = new EdgeUpdateFeedbackApprovalRequest()
        {
            FeedbackId = "feedbackId",
            ApprovalStatus = ApprovalStatus.Approved
        };

        EdgeFeedbackResult feedback = lwClient.UpdateFeedbackApproval(edgeUpdateFeedbackApprovalRequest);

        Console.WriteLine(feedback.ApprovalState);
    }

    public static void DeleteFeedback()
    {
        using var lwClient = new EdgeLanguageWeaverClient().Build();

        EdgeDeleteFeedbackRequest deleteFeedbackRequest = new EdgeDeleteFeedbackRequest()
        {
            FeedbackId = "feedbackId"
        };

        lwClient.DeleteFeedback(deleteFeedbackRequest);
    }

    public static void RetrieveFeedback()
    {
        using var lwClient = new EdgeLanguageWeaverClient().Build();

        int pageNumber = 1;
        int pageSize = 50;
        EdgeFeedbackListResult edgeFeedbackList;
        do
        {
            edgeFeedbackList = lwClient.GetEdgeFeedback(pageNumber, pageSize);
            //process feedback
            List<EdgeFeedbackResult> feedbackList = edgeFeedbackList.FeedbackList;
            feedbackList.Take(3)
                .ToList()
                .ForEach(edgeFeedbackResult => Console.WriteLine(edgeFeedbackResult.FeedbackId));
            pageNumber++;
        } while (edgeFeedbackList.FeedbackList.Count == pageSize);
    }

    public static void FilterFeedback()
    {
        using var lwClient = new EdgeLanguageWeaverClient().Build();

        EdgeFilterFeedbackRequest filterFeedbackRequest = new EdgeFilterFeedbackRequest()
        {
            LanguagePairId = "languagePairId",
            SourceText = "sourceText",
            MachineTranslation = "machineTranslation",
            SuggestedTranslation = "suggestedTranslation",
            Reviewer = "reviewer username",
            User = "creation username",
            ApprovalStatus = ApprovalStatus.Approved,
            StartDate = "10/10/2021",
            EndDate = "10/10/2022",
        };

        int pageNumber = 1;
        EdgeFeedbackListResult filteredFeedback;
        do
        {
            filteredFeedback = lwClient.GetEdgeFeedback(filterFeedbackRequest, pageNumber);
            //process feedback
            List<EdgeFeedbackResult> feedbackList = filteredFeedback.FeedbackList;
            feedbackList.Take(3)
                .ToList()
                .ForEach(edgeFeedbackResult => Console.WriteLine(edgeFeedbackResult.FeedbackId));
            pageNumber++;
        } while (filteredFeedback.FeedbackList.Count > 0);
    }
}